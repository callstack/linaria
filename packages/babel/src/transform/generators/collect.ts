import type { PluginItem } from '@babel/core';

import { buildOptions } from '@linaria/utils';

import { filename as collectorPlugin } from '../../plugins/collector';
import { getLinariaMetadata } from '../../utils/withLinariaMetadata';
import type { ICollectAction, SyncScenarioForAction } from '../types';

/**
 * Parses the specified file, finds tags, applies run-time replacements,
 * removes dead code.
 */
// eslint-disable-next-line require-yield
export function* collect(
  this: ICollectAction
): SyncScenarioForAction<ICollectAction> {
  const { babel, options } = this.services;
  const { valueCache } = this.data;
  const { entrypoint } = this;
  const { loadedAndParsed, name } = entrypoint;

  if (loadedAndParsed.evaluator === 'ignored') {
    throw new Error('entrypoint was ignored');
  }

  const transformPlugins: PluginItem[] = [
    [
      collectorPlugin,
      {
        ...options.pluginOptions,
        values: valueCache,
      },
    ],
  ];

  const transformConfig = buildOptions({
    envName: 'linaria',
    plugins: transformPlugins,
    sourceMaps: true,
    sourceFileName: name,
    inputSourceMap: options.inputSourceMap,
    root: options.root,
    ast: true,
    babelrc: false,
    configFile: false,
    sourceType: 'unambiguous',
  });

  const result = babel.transformFromAstSync(
    loadedAndParsed.ast,
    loadedAndParsed.code,
    {
      ...transformConfig,
      cwd: options.root,
      filename: name,
    }
  );

  if (!result || !result.ast?.program) {
    throw new Error('Babel transform failed');
  }

  const linariaMetadata = getLinariaMetadata(result.metadata);

  return {
    ast: result.ast,
    code: result.code,
    map: result.map,
    metadata: linariaMetadata ?? null,
  };
}
