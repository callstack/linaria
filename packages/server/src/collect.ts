import type { AtRule, ChildNode } from 'postcss';
import postcss from 'postcss';

type CollectResult = {
  critical: string;
  other: string;
};

interface ClassnameModifiers {
  ignoredClasses?: string[];
  blockedClasses?: string[];
}

/**
 * Used to escape `RegExp`
 * [syntax characters](https://262.ecma-international.org/7.0/#sec-regular-expressions-patterns).
 */
function escapeRegex(string: string) {
  return string.replace(/[\\^$.*+?()[\]{}|]/g, '\\$&');
}

const extractClassesFromHtml = (
  html: string,
  ignoredClasses: string[]
): RegExp => {
  const htmlClasses: string[] = [];
  const regex = /\s+class="([^"]+)"/gm;
  let match = regex.exec(html);
  const ignoredClassesDeduped = new Set(ignoredClasses);

  while (match !== null) {
    match[1].split(' ').forEach((className) => {
      // eslint-disable-next-line no-param-reassign
      className = escapeRegex(className);
      if (className !== '' && !ignoredClassesDeduped.has(className)) {
        htmlClasses.push(className);
      }
    });
    match = regex.exec(html);
  }

  return new RegExp(htmlClasses.join('|'), 'gm');
};

/**
 * This utility extracts critical CSS from given HTML and CSS file to be used in SSR environments
 * @param {string} html the HTML from which classes will be parsed
 * @param {string} css the CSS file from which selectors will be parsed and determined as critical or other
 * @param {string[]} ignoredClasses classes that, when present in the HTML, will not be included in the regular expression used to match selectors
 * @param {string[]} blockedClasses classes that, when contained in a selector, will cause the selector to be marked as not critical
 * @returns {CollectResult} object containing the critical and other CSS styles
 */
export default function collect(
  html: string,
  css: string,
  classnameModifiers?: ClassnameModifiers
): CollectResult {
  const animations = new Set();
  const other = postcss.root();
  const critical = postcss.root();
  const stylesheet = postcss.parse(css);
  const ignoredClasses = classnameModifiers?.ignoredClasses ?? [];
  const blockedClasses = classnameModifiers?.blockedClasses ?? [];

  const htmlClassesRegExp = extractClassesFromHtml(html, ignoredClasses);
  const blockedClassesSanitized = blockedClasses.map(escapeRegex);
  const blockedClassesRegExp = new RegExp(
    blockedClassesSanitized.join('|'),
    'gm'
  );

  const isCritical = (rule: ChildNode) => {
    // Only check class names selectors
    if ('selector' in rule && rule.selector.startsWith('.')) {
      const isExcluded =
        blockedClasses.length > 0 && blockedClassesRegExp.test(rule.selector);
      if (isExcluded) return false;

      return Boolean(rule.selector.match(htmlClassesRegExp));
    }

    return true;
  };

  const handleAtRule = (rule: AtRule) => {
    if (rule.name === 'keyframes') {
      return;
    }

    const criticalRule = rule.clone();
    const otherRule = rule.clone();

    let removedNodesFromOther = 0;
    criticalRule.each((childRule: ChildNode, index: number) => {
      if (isCritical(childRule)) {
        otherRule.nodes[index - removedNodesFromOther]?.remove();
        removedNodesFromOther += 1;
      } else {
        childRule.remove();
      }
    });

    rule.remove();

    if (criticalRule.nodes.length > 0) {
      critical.append(criticalRule);
    }
    if (otherRule.nodes.length > 0) {
      other.append(otherRule);
    }
  };

  stylesheet.walkAtRules('font-face', (rule) => {
    /**
     * @font-face rules may be defined also in CSS conditional groups (eg. @media)
     * we want only handle those from top-level, rest will be handled in stylesheet.walkRules
     */
    if (rule.parent?.type === 'root') {
      critical.append(rule);
    }
  });

  const walkedAtRules = new Set();

  stylesheet.walkRules((rule) => {
    if (
      rule.parent &&
      'name' in rule.parent &&
      (rule.parent as { name: string }).name === 'keyframes'
    ) {
      return;
    }

    if (rule.parent?.type === 'atrule') {
      if (!walkedAtRules.has(rule.parent)) {
        handleAtRule(rule.parent as AtRule);
        walkedAtRules.add(rule.parent);
      }
      return;
    }

    if (isCritical(rule)) {
      critical.append(rule);
    } else {
      other.append(rule);
    }
  });

  critical.walkDecls(/animation/, (decl) => {
    animations.add(decl.value.split(' ')[0]);
  });

  stylesheet.walkAtRules('keyframes', (rule) => {
    if (animations.has(rule.params)) {
      critical.append(rule);
    }
  });

  return {
    critical: critical.toString(),
    other: other.toString(),
  };
}
