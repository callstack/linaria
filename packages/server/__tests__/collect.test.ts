import dedent from 'dedent';
import prettier from 'prettier';
import collect from '../src/collect';

const prettyPrint = (src: string) => prettier.format(src, { parser: 'scss' });

const testCollect = (html: string, css: string) => {
  const { critical, other } = collect(html, css);
  test('critical', () => expect(prettyPrint(critical)).toMatchSnapshot());
  test('other', () => expect(prettyPrint(other)).toMatchSnapshot());
};

const html = dedent`
  <div class="linaria lily">
    <span class="lotus"></div>
  </div>
`;

describe('collects complex css', () => {
  const css = dedent`
    .lotus {
      vertical-align: top;
    }

    @media (max-width: 1200px) {
      .lotus {
        vertical-align: bottom;
      }
    }

    @supports (object-fit: cover) {
      .unrelated-nested {
        float: left;
        animation: custom-animation;
      }

      .unrelated-nested2 {
        float: left;
      }
    }

    @supports (object-fit: contain) {
      .lotus {
        object-fit: contain;
      }

      .linaria::before,
      .linaria::after {
        content: '';
        object-fit: contain;
      }
    }

    @keyframes custom-animation {
      0% { opacity: 0 }
      50% { opacity: 0 }
      100% { opacity: 1 }
    }

    .linaria {
      float: left;
      flex: 1;
      animation: custom-animation .2s;
    }

    .linaria:hover {
      float: right;
    }

    .linaria > span,
    .linaria + .linaria,
    .linaria ~ div {
      display: none;
    }

    .linaria > span {
      display: none;
    }

    .linaria::after {
      display: block;
    }

    .unrelated {
      animation-name: custom-animation;
    }

    .unrelated2 {
      animation: custom-animation .3s;
    }

    .lily {
      color: #fff;
    }

    [data-theme=dark] .lily {
      color: #000;
    }

    .unrelated3 {
      flex: 0;
    }

    .linaria ~ div {}
    .linaria.linaria2{}
  `;

  testCollect(html, css);
});

describe('simple class name', () => {
  const css = dedent`
    .linaria {}
    .classname {}
  `;

  testCollect(html, css);
});

describe('classname in @rule', () => {
  const css = dedent`
    @supports (object-fit: cover) { .linaria {} }
    @media (min-width: 600px) { .linaria {} }
    @charset () { .linaria {} }
    @import () { .linaria {} }
    @namespace () { .linaria {} }
    @media () { .linaria {} }
    @supports () { .linaria {} }
    @document () { .linaria {} }
    @page () { .linaria {} }
    @keyframes () { .linaria {} }
    @viewport () { .linaria {} }
    @counter-style () { .linaria {} }
    @font-feature-values () { .linaria {} }

    @supports (object-fit: cover) { .other {} }
    @media (min-width: 600px) { .other {} }
    @charset () { .other {} }
    @import () { .other {} }
    @namespace () { .other {} }
    @media () { .other {} }
    @supports () { .other {} }
    @document () { .other {} }
    @page () { .other {} }
    @keyframes () { .other {} }
    @viewport () { .other {} }
    @counter-style () { .other {} }
    @font-feature-values () { .other {} }
  `;

  testCollect(html, css);
});

describe('works with CSS combinators', () => {
  const css = dedent`
    .linaria + span {}
    .linaria ~ div {}
    .linaria > a {}
    .linaria b {}

    .other + span {}
    .other ~ div {}
    .other > a {}
    .other b {}
  `;
  testCollect(html, css);
});

describe('works with pseudo-class and pseudo-elements', () => {
  const css = dedent`
    .linaria:active {}
    .linaria::before {}

    .other:active {}
    .other::before {}
  `;
  testCollect(html, css);
});

describe('works with global css', () => {
  const css = dedent`
    body { font-size: 13.37px; }

    html { -webkit-font-smoothing: antialiased; }

    h1 { font-weight: bold; }

    .linaria:active {}
    .linaria::before {}

    .other:active {}
    .other::before {}
  `;

  const { critical, other } = collect(html, css);

  test('critical', () => expect(prettyPrint(critical)).toMatchSnapshot());
  test('other', () => expect(prettyPrint(other)).toMatchSnapshot());
});

describe('handles top-level @font-face', () => {
  const css = dedent`
    @font-face {
      font-family: MyFont;
      font-weight: normal;
      font-style: normal;
      src: url(MyFont.woff);
    }
  `;
  const { critical, other } = collect(html, css);

  test('critical', () => expect(prettyPrint(critical)).toMatchSnapshot());
  test('other', () => expect(prettyPrint(other)).toMatchSnapshot());
});

// there was a bug when the whole atrule was included for each child rule
describe('include atrule once', () => {
  const css = dedent`
    @media screen {
      body {
        font-size: 10px;
      }
      h1 {
        font-size: 20px;
      }
      .class {
        font-size: 15px;
      }
    }
  `;
  const { critical, other } = collect(html, css);

  test('critical', () => expect(prettyPrint(critical)).toMatchSnapshot());
  test('other', () => expect(prettyPrint(other)).toMatchSnapshot());
});

describe('ignore empty class attribute', () => {
  const html = dedent`
    <div class=""></div>
    <div class="a"></div>
    <div class=""></div>
    <div class="b"></div>
    <div class=""></div>
  `;

  const css = dedent`
    .not-exist {}
  `;

  const { critical } = collect(html, css);
  test('critical should be empty', () => expect(critical).toEqual(''));
});
