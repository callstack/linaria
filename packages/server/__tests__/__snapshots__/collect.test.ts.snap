// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`classname in @rule critical 1`] = `
"@supports (object-fit: cover) {
  .linaria {
  }
}
@media (min-width: 600px) {
  .linaria {
  }
}
@charset() {
  .linaria {
  }
}
@import() {
  .linaria {
  }
}
@namespace () {
  .linaria {
  }
}
@media() {
  .linaria {
  }
}
@supports () {
  .linaria {
  }
}
@document() {
  .linaria {
  }
}
@page() {
  .linaria {
  }
}
@viewport() {
  .linaria {
  }
}
@counter-style() {
  .linaria {
  }
}
@font-feature-values() {
  .linaria {
  }
}
"
`;

exports[`classname in @rule other 1`] = `
"@supports (object-fit: cover) {
  .other {
  }
}
@media (min-width: 600px) {
  .other {
  }
}
@charset() {
  .other {
  }
}
@import() {
  .other {
  }
}
@namespace () {
  .other {
  }
}
@media() {
  .other {
  }
}
@supports () {
  .other {
  }
}
@document() {
  .other {
  }
}
@page() {
  .other {
  }
}
@viewport() {
  .other {
  }
}
@counter-style() {
  .other {
  }
}
@font-feature-values() {
  .other {
  }
}
"
`;

exports[`collects complex css critical 1`] = `
".lotus {
  vertical-align: top;
}
@media (max-width: 1200px) {
  .lotus {
    vertical-align: bottom;
  }
}
@supports (object-fit: contain) {
  .lotus {
    object-fit: contain;
  }

  .linaria::before,
  .linaria::after {
    content: \\"\\";
    object-fit: contain;
  }
}
.linaria {
  float: left;
  flex: 1;
  animation: custom-animation 0.2s;
}
.linaria:hover {
  float: right;
}
.linaria > span,
.linaria + .linaria,
.linaria ~ div {
  display: none;
}
.linaria > span {
  display: none;
}
.linaria::after {
  display: block;
}
.lily {
  color: #fff;
}
[data-theme=\\"dark\\"] .lily {
  color: #000;
}
.linaria ~ div {
}
.linaria.linaria2 {
}
@keyframes custom-animation {
  0% {
    opacity: 0;
  }
  50% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}
"
`;

exports[`collects complex css other 1`] = `
"@supports (object-fit: cover) {
  .unrelated-nested {
    float: left;
    animation: custom-animation;
  }

  .unrelated-nested2 {
    float: left;
  }
}

.unrelated {
  animation-name: custom-animation;
}

.unrelated2 {
  animation: custom-animation 0.3s;
}

.unrelated3 {
  flex: 0;
}
"
`;

exports[`handles top-level @font-face critical 1`] = `
"@font-face {
  font-family: MyFont;
  font-weight: normal;
  font-style: normal;
  src: url(MyFont.woff);
}
"
`;

exports[`handles top-level @font-face other 1`] = `""`;

exports[`include atrule once critical 1`] = `
"@media screen {
  body {
    font-size: 10px;
  }
  h1 {
    font-size: 20px;
  }
}
"
`;

exports[`include atrule once other 1`] = `
"@media screen {
  .class {
    font-size: 15px;
  }
}
"
`;

exports[`simple class name critical 1`] = `
".linaria {
}
"
`;

exports[`simple class name other 1`] = `
".classname {
}
"
`;

exports[`works with CSS combinators critical 1`] = `
".linaria + span {
}
.linaria ~ div {
}
.linaria > a {
}
.linaria b {
}
"
`;

exports[`works with CSS combinators other 1`] = `
".other + span {
}
.other ~ div {
}
.other > a {
}
.other b {
}
"
`;

exports[`works with global css critical 1`] = `
"body {
  font-size: 13.37px;
}
html {
  -webkit-font-smoothing: antialiased;
}
h1 {
  font-weight: bold;
}
.linaria:active {
}
.linaria::before {
}
"
`;

exports[`works with global css other 1`] = `
".other:active {
}
.other::before {
}
"
`;

exports[`works with pseudo-class and pseudo-elements critical 1`] = `
".linaria:active {
}
.linaria::before {
}
"
`;

exports[`works with pseudo-class and pseudo-elements other 1`] = `
".other:active {
}
.other::before {
}
"
`;
